libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,177 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diskqueue
+
+import (
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/feature"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/paths"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+// Settings contains the configuration fields to create a new disk queue
+// or open an existing one.
+type Settings struct {
+	// The path on disk of the queue's main metadata file.
+	// Paths for data segment files are created by appending ".{segmentIndex}" to
+	// this path.
+	// If blank, the default is "queue.dat" within the beat's data directory.
+	Path string
+
+	// The size in bytes of one data page in the on-disk buffer. To minimize
+	// data loss if there is an error, this should match the page size of the
+	// target filesystem.
+	PageSize uint32
+
+	// MaxBufferSize is the maximum number of bytes that the queue should
+	// ever occupy on disk. A value of 0 means the queue can grow until the
+	// disk is full.
+	MaxBufferSize uint64
+
+	// The destination for log messages related to the disk queue.
+	Logger *logp.Logger
```

Comment:
We need this under 'Settings', or can it be another parameter to the constructor? For 'generic support' types I'd prefer to have them at the top (bikeshedding :) ).

Reply:
`Settings` is the aggregate type that the constructor accepts (`func NewQueue(settings Settings) (queue.Queue, error) ...`) -- if the logger was a separate parameter it would still have to be put right back in the queue structure, and I wanted to have all input parameters (as opposed to logical / derived) in the same struct for clarity.

I could move `Logger` to the top of the `Settings` struct definition though if that's what you mean in the second half?

Reply:
> I could move Logger to the top of the Settings struct definition though if that's what you mean in the second half?

Yes :)